### ISSUE TITLE
Implement a proper deep (Lorri-style) autodetection

### ISSUE THREAD

**Issue by shajra** [shajra/direnv-nix-lorelei#1]
To address performance, I removed a trace that was slowing evaluation times down (by patching Lorri's code in a Nix expression).  But now I realize that this affects the auto-detection of files to watch.

I kind of like the idea of having both a shallow and a deep auto-detection, letting the user chose which they prefer.

The documentation of Lorelei currently states that auto-detection misses some files.  I _think_ that's still right, but I need to implement the deep algorithm (not patch Lorri) to validate that statement.  It's possible my documentation is wrong in that regard.

**Comment by shajra** [comment]
Okay, I ran a small test, and without the patch, auto-detection does pick up a few more files.  In my case, it picked up that my project's Niv `sources.json` file that was read.

But a 2x slowdown seems like a heavy price to pay for a few JSON files.  And it still doesn't pick up on files I'm more interested in watching, like the Cabal files in my Haskell projects.

Still, I'm inclined to implement the deep auto-detection still... just for completeness.  So that people don't feel I pushed that decision on them.  I'm just tired of there being so many of these scripts.  I think it would be good if there was one that really did subsume them all.
